File: LeetCode 热题 HOT 100/二叉树/116. 填充每个节点的下一个右侧节点指针.py
# 给定一个 完美二叉树 ，其所有叶子节点都在同一层，每个父节点都有两个子节点。二叉树定义如下：

# struct Node {
#   int val;
#   Node * left;
#   Node * right;
#   Node * next;}
# 填充它的每个 next 指针，让这个指针指向其下一个右侧节点。如果找不到下一个右侧节点，则将 next 指针设置为 NULL。

# 初始状态下，所有 next 指针都被设置为 NULL。

# 示例 1：
# 输入：root = [1, 2, 3, 4, 5, 6, 7]
# 输出：[1,  # ,2,3,#,4,5,6,7,#]
# 解释：给定二叉树如图 A 所示，你的函数应该填充它的每个 next 指针，以指向其下一个右侧节点，
# 如图 B 所示。序列化的输出按层序遍历排列，同一层节点由 next 指针连接，'#' 标志着每一层的结束。

# 示例 2:
# 输入：root= []
# 输出：[]


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


from collections import deque


class Solution:
    def connect(root):
        if not root:
            return None
        points = deque([root])

        def nextDir(node):
            if node.left:
                points.append(node.left)
            if node.right:
                points.append(node.right)

        while points:
            size = len(points)
            for i in range(size):
                node = points.popleft()
                nextDir(node)
                if i == size - 1:
                    break
                node.next = points[0]
        return root

File: LeetCode 热题 HOT 100/二叉树/116. 填充每个节点的下一个右侧节点指针.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function (root) {
  if (root === null) {
    return null
  }

  points = []
  points.push(root)

  function nextDir(node) {
    if (node.left) {
      points.push(node.left)
    }
    if (node.right) {
      points.push(node.right)
    }
  }

  while (points.length !== 0) {
    size = points.length
    for (let index = 0; index < size; index++) {
      point = points.shift()
      nextDir(point)
      if (index === size - 1) {
        break
      }

      point.next = points[0]
    }
  }
  return root
};